File: Xero.WebApi/Controllers/RunExecutionsController.cs
using Microsoft.AspNetCore.Mvc;
using Xero.WebApi.Data;
using Xero.WebApi.Models;
using Xero.WebApi.Services;

namespace Xero.WebApi.Controllers;

/// <summary>
/// Manages NRT run executions — trigger ad-hoc runs and query execution history.
/// </summary>
[ApiController]
[Route("api/run-executions")]
[Produces("application/json")]
public sealed class RunExecutionsController : ControllerBase
{
    private readonly INrtService           _nrtService;
    private readonly RunExecutionRepository _runRepository;

    public RunExecutionsController(INrtService nrtService, RunExecutionRepository runRepository)
    {
        _nrtService    = nrtService;
        _runRepository = runRepository;
    }

    /// <summary>
    /// Executes a full NRT pipeline: load reference + target data, compare, and save results.
    /// The call is synchronous and returns once all steps complete.
    /// Optionally pass a <c>DefinitionId</c> to run from a saved definition (only <c>ValuationDate</c> is required from the body in that case).
    /// </summary>
    /// <response code="200">Run completed — inspect <c>Passed</c> and diff counts.</response>
    /// <response code="400">Request body is invalid or required fields are missing.</response>
    /// <response code="500">An unexpected error occurred during the run.</response>
    [HttpPost]
    [ProducesResponseType(typeof(RunExecutionResponse), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status400BadRequest)]
    [ProducesResponseType(StatusCodes.Status500InternalServerError)]
    public async Task<ActionResult<RunExecutionResponse>> ExecuteRun(
        [FromBody] NrtRunRequest request,
        CancellationToken ct)
    {
        // When running ad-hoc (no definition), validate the connection strings are present
        if (!request.DefinitionId.HasValue)
        {
            if (string.IsNullOrWhiteSpace(request.Reference.ConnectionString))
                return BadRequest("Reference.ConnectionString is required.");

            if (string.IsNullOrWhiteSpace(request.Reference.Query))
                return BadRequest("Reference.Query is required.");

            if (string.IsNullOrWhiteSpace(request.Target.ConnectionString))
                return BadRequest("Target.ConnectionString is required.");

            if (string.IsNullOrWhiteSpace(request.Target.Query))
                return BadRequest("Target.Query is required.");
        }

        var result = await _nrtService.ExecuteRunAsync(request, ct);
        return Ok(result);
    }

    /// <summary>
    /// Returns a paginated list of past NRT run executions ordered by most recent first.
    /// </summary>
    [HttpGet]
    [ProducesResponseType(typeof(IReadOnlyList<RunExecutionSummary>), StatusCodes.Status200OK)]
    public async Task<ActionResult<IReadOnlyList<RunExecutionSummary>>> GetRuns(
        [FromQuery] int page     = 1,
        [FromQuery] int pageSize = 20,
        CancellationToken ct     = default)
    {
        if (page < 1)       page     = 1;
        if (pageSize < 1)   pageSize = 20;
        if (pageSize > 200) pageSize = 200;

        var runs = await _runRepository.GetRunsAsync(page, pageSize, ct);
        return Ok(runs);
    }

    /// <summary>
    /// Returns the summary for a single NRT run execution.
    /// </summary>
    [HttpGet("{runId:int}")]
    [ProducesResponseType(typeof(RunExecutionSummary), StatusCodes.Status200OK)]
    [ProducesResponseType(StatusCodes.Status404NotFound)]
    public async Task<ActionResult<RunExecutionSummary>> GetRun(int runId, CancellationToken ct)
    {
        var run = await _runRepository.GetRunAsync(runId, ct);
        if (run is null)
            return NotFound($"Run execution {runId} not found.");
        return Ok(run);
    }
}

File: Xero.JavaWebApi/src/main/java/com/xero/webapi/controller/RunExecutionsController.java
package com.xero.webapi.controller;

import com.xero.webapi.model.DbSettingsDto;
import com.xero.webapi.model.NrtRunRequest;
import com.xero.webapi.model.PagedResult;
import com.xero.webapi.model.RunExecutionResponse;
import com.xero.webapi.model.RunExecutionSummary;
import com.xero.webapi.repository.RunExecutionRepository;
import com.xero.webapi.service.NrtService;
import jakarta.validation.Valid;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import java.util.Map;

/**
 * Lower-level execution surface — accepts either an inline {@link NrtRunRequest} or a request that
 * carries only a {@code DefinitionId} (in which case the service merges the saved definition).
 * Mirrors the .NET {@code RunExecutionsController}.
 */
@RestController
@RequestMapping("/api/run-executions")
public class RunExecutionsController {

    private final NrtService             nrtService;
    private final RunExecutionRepository runRepository;

    public RunExecutionsController(NrtService nrtService, RunExecutionRepository runRepository) {
        this.nrtService    = nrtService;
        this.runRepository = runRepository;
    }

    @PostMapping
    public ResponseEntity<?> executeRun(@Valid @RequestBody NrtRunRequest request) {
        // Skip ref/tgt validation when a DefinitionId is supplied — the service merges the
        // connection settings from the persisted definition.
        if (request.getDefinitionId() == null) {
            String missing = validateConnections(request);
            if (missing != null) {
                return ResponseEntity.badRequest().body(Map.of("error", missing));
            }
        }
        RunExecutionResponse response = nrtService.executeRun(request);
        return ResponseEntity.ok(response);
    }

    @GetMapping
    public PagedResult<RunExecutionSummary> listRuns(@RequestParam(defaultValue = "1") int page,
                                                     @RequestParam(defaultValue = "20") int pageSize) {
        page     = Math.max(1, page);
        pageSize = Math.min(200, Math.max(1, pageSize));
        var items = runRepository.getRuns(page, pageSize);
        int total = runRepository.count();
        return new PagedResult<>(items, total, page, pageSize);
    }

    @GetMapping("/{runId}")
    public ResponseEntity<?> getRun(@PathVariable int runId) {
        RunExecutionSummary summary = runRepository.getRun(runId);
        if (summary == null) {
            return ResponseEntity.status(404)
                    .body(Map.of("error", "Run execution " + runId + " not found."));
        }
        return ResponseEntity.ok(summary);
    }

    private static String validateConnections(NrtRunRequest request) {
        DbSettingsDto ref = request.getReference();
        DbSettingsDto tgt = request.getTarget();
        if (ref == null || isBlank(ref.getConnectionString())) {
            return "Reference.ConnectionString is required.";
        }
        if (isBlank(ref.getQuery())) {
            return "Reference.Query is required.";
        }
        if (tgt == null || isBlank(tgt.getConnectionString())) {
            return "Target.ConnectionString is required.";
        }
        if (isBlank(tgt.getQuery())) {
            return "Target.Query is required.";
        }
        return null;
    }

    private static boolean isBlank(String s) {
        return s == null || s.isBlank();
    }
}
